037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state.
+The cluster operator logs a warning about the rebalancing not happened.
+It's the user who should deal with this scenario:
+
+* Checking the reason why the `KafkaRebalance` is in the `NotReady` state and deleting the custom resource.
+* If the user still wants rebalancing, they have to create a new `KafkaRebalance` custom resource manually.
+
+### Auto-rebalancing on scaling down
+
+1. The user scales down the cluster by decreasing the `spec.kafka.replicas` field.
+2. The `KafkaReconciler` takes care of it and, if the `spec.kafka.autoRebalance` contains an entry with `remove-brokers` mode, it creates a `KafkaRebalance` custom resource named as `auto-<cluster-name>-rebalance` by using the corresponding template and sets:
+   * the `spec.mode` as `remove-brokers` and the `spec.brokers` with the list of brokers that will be removed.
+   * the `spec.autoApproval` from the template as described by this [proposal](https://github.com/strimzi/proposals/pull/56).
+3. The `KafkaRebalanceAssemblyOperator` detects the new rebalancing resource and acts as usual by requesting Cruise Control to generate an optimization proposal via the REST API on the `/remove_broker` endpoint.
+4. When the optimization proposal is ready, the `KafkaRebalanceAssemblyOperator` checks the auto-approval:
+   * If criteria are met or they are not specified at all, it approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.
+   * If criteria are not met, the `KafkaRebalance` custom resource status will move to `NotReady`, using a warning condition to describe why it was not approved automatically.
+5. If/When the rebalancing ends successfully, the corresponding `KafkaRebalance` custom resource is deleted by the `KafkaReconciler`.
+6. Finally, the `KafkaReconciler` can continue to scale down the cluster by deleting the brokers.
+
+#### Error handling
+
+During the above flow, errors can happen before the scaling down:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state.
+The cluster operator logs a warning about the rebalancing not happened.
+
+The scale down cannot proceed because of the brokers to be removed are still hosting some replicas.
+This behavior is related to this [proposal](https://github.com/strimzi/proposals/pull/54).
+
+On the next reconciliation, the cluster operator will try to do a scale down again, so running the flow as before.
+The previous `KafkaRebalance` custom resource in the `NotReady` state is overridden to generate a new optimization proposal.
```

Comment:
> you wait for 5 minutes and timeout in reconciliation no 1.

When do you mean it times out? When doing the rebalancing or the scale down?
The rebalancing cannot time out because it happens across more reconciliation of course. The KRAO has a different internal periodic pool to check the status through the CC REST API. If at some point it fails and the `KafkaRebalance` moves to `NotReady`, the scale down doesn't start.
If it times out during scale down, it means rebalancing went ok and the `KafkaRebalance` was deleted. Anyway what I see is that on next reconcile due to the auto-rebalancing enabled, the operator could try to restart a rebalance again (creating the `KafkaRebalance`) so we would need to keep track of the rebalancing already happened? Is this what you mean?